Code/Tools/AssetProcessor/native/unittests/AssetProcessingStateDataUnitTests.cpp

```diff
@@ -595,123 +172,793 @@ namespace AssetProcessor
+        EXPECT_FALSE(m_connection.GetScanFolderByProductID(0, scanFolder));
+        EXPECT_FALSE(m_connection.GetScanFolderByPortableKey("sadfsadfsadfsadfs", scanFolder));
+        scanFolders.clear();
+
+        AddDefaultScanFolder();
+
+        // Get all scan folders, there should be one we just added
+        scanFolders.clear();
+        EXPECT_TRUE(m_connection.GetScanFolders(scanFolders));
+        EXPECT_EQ(scanFolders.size(), 1);
+        EXPECT_TRUE(ScanFoldersContainScanPath(scanFolders, "c:/O3DE/dev"));
+        EXPECT_TRUE(ScanFoldersContainScanFolderID(scanFolders, m_defaultScanFolder.m_scanFolderID));
+        EXPECT_TRUE(ScanFoldersContainPortableKey(scanFolders, m_defaultScanFolder.m_portableKey.c_str()));
+        EXPECT_TRUE(ScanFoldersContainPortableKey(scanFolders, "rootportkey"));
+    }
+
+    TEST_F(AssetProcessingStateDataUnitTest, TestScanFolder_AddDuplicateScanFolder_GetsSameId)
+    {
+        AddDefaultScanFolder();
+
+        // Add the same folder again, should not add another because it already exists, so we should get the same id
+        // not only that, but the path should update.
+        ScanFolderDatabaseEntry dupeScanFolder("c:/O3DE/dev1", "dev", "rootportkey");
+        dupeScanFolder.m_scanFolderID = AzToolsFramework::AssetDatabase::InvalidEntryId;
+        EXPECT_TRUE(m_connection.SetScanFolder(dupeScanFolder));
+        EXPECT_EQ(dupeScanFolder, m_defaultScanFolder) << "AssetProcessingStateDataTest Failed - scan folder failed to add";
+
+        EXPECT_EQ(dupeScanFolder.m_portableKey, m_defaultScanFolder.m_portableKey);
+        EXPECT_EQ(dupeScanFolder.m_scanFolderID, m_defaultScanFolder.m_scanFolderID);
+
+        // Get all scan folders, there should only the one we added
+        ScanFolderDatabaseEntryContainer scanFolders;
+        EXPECT_TRUE(m_connection.GetScanFolders(scanFolders));
+        EXPECT_EQ(scanFolders.size(), 1);
+        EXPECT_TRUE(ScanFoldersContainScanPath(scanFolders, "c:/O3DE/dev1"));
+        EXPECT_TRUE(ScanFoldersContainScanFolderID(scanFolders, m_defaultScanFolder.m_scanFolderID));
+        EXPECT_TRUE(ScanFoldersContainPortableKey(scanFolders, m_defaultScanFolder.m_portableKey.c_str()));
+        EXPECT_TRUE(ScanFoldersContainPortableKey(scanFolders, "rootportkey"));
+    }
+
+    TEST_F(AssetProcessingStateDataUnitTest, TestScanFolder_RetrieveScanFolderById_Succeeds)
+    {
+        AddDefaultScanFolder();
+
+        // Retrieve the one we just made by id
+        ScanFolderDatabaseEntry retrieveScanfolderById;
+        EXPECT_TRUE(m_connection.GetScanFolderByScanFolderID(m_defaultScanFolder.m_scanFolderID, retrieveScanfolderById));
+        EXPECT_NE(retrieveScanfolderById.m_scanFolderID, AzToolsFramework::AssetDatabase::InvalidEntryId) << "AssetProcessingStateDataTest Failed - scan folder failed to add";
+        EXPECT_EQ(retrieveScanfolderById.m_scanFolderID, m_defaultScanFolder.m_scanFolderID) << "AssetProcessingStateDataTest Failed - scan folder failed to add";
```

Comment:
These error messages aren't super useful, the same message "scan folder failed to add" is repeated everywhere. They could probably just all be removed.